cpp/src/parquet/metadata.cc

```diff
@@ -132,6 +132,31 @@ std::shared_ptr<Statistics> MakeColumnStats(const format::ColumnMetaData& meta_d
   throw ParquetException("Can't decode page statistics for selected column type");
 }
 
+template <typename Metadata>
+std::shared_ptr<KeyValueMetadata> CopyKeyValueMetadata(const Metadata& source) {
+  std::shared_ptr<KeyValueMetadata> metadata = nullptr;
+  if (source.__isset.key_value_metadata) {
+    metadata = std::make_shared<KeyValueMetadata>();
+    for (const auto& it : source.key_value_metadata) {
+      metadata->Append(it.key, it.value);
+    }
+  }
+  return metadata;
+}
+
+template <typename Metadata>
+void SetKeyValueMetadata(Metadata& metadata, const KeyValueMetadata& source) {
+  metadata.key_value_metadata.clear();
+  metadata.key_value_metadata.reserve(static_cast<size_t>(source.size()));
+  for (int64_t i = 0; i < source.size(); ++i) {
+    format::KeyValue kv_pair;
+    kv_pair.__set_key(source.key(i));
+    kv_pair.__set_value(source.value(i));
+    metadata.key_value_metadata.push_back(kv_pair);
+  }
+  metadata.__isset.key_value_metadata = true;
```

Comment:
Can we use `void __set_key_value_metadata(const std::vector<KeyValue> & val);` instead?